apps/agent/contracts/Agent.sol

```diff
@@ -0,0 +1,147 @@
+            returndatacopy(ptr, 0, size)
+
+            // revert instead of invalid() bc if the underlying call failed with invalid() it already wasted gas.
+            // if the call returned error data, forward it
+            switch result case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+
+    /**
+    * @notice Set `_designatedSigner` as the designated signer of the app, which will be able to sign messages on behalf of the app
+    * @param _designatedSigner Address that will be able to sign messages on behalf of the app
+    */
+    function setDesignatedSigner(address _designatedSigner)
+        external
+        authP(DESIGNATE_SIGNER_ROLE, arr(_designatedSigner))
+    {
+        // Prevent an infinite loop by setting the app itself as its designated signer.
+        // An undetectable loop can be created by setting a different contract as the
+        // designated signer which calls back into `isValidSignature`.
+        // Given that `isValidSignature` is always called with just 50k gas, the max
+        // damage of the loop is wasting 50k gas.
+        require(_designatedSigner != address(this), ERROR_DESIGNATED_TO_SELF);
+
+        address oldDesignatedSigner = designatedSigner;
+        designatedSigner = _designatedSigner;
+
+        emit SetDesignatedSigner(msg.sender, oldDesignatedSigner, _designatedSigner);
+    }
+
+    /**
+    * @notice Pre-sign hash `_hash`
+    * @param _hash Hash that will be considered signed regardless of the signature checked with 'isValidSignature()'
+    */
+    function presignHash(bytes32 _hash)
+        external
+        authP(ADD_PRESIGNED_HASH_ROLE, arr(_hash))
+    {
+        isPresigned[_hash] = true;
+
+        emit PresignHash(msg.sender, _hash);
+    }
+
+    function isForwarder() external pure returns (bool) {
+        return true;
+    }
+
+    function supportsInterface(bytes4 interfaceId) external pure returns (bool) {
+        return
+            interfaceId == ERC1271_INTERFACE_ID ||
+            interfaceId == ERC165_INTERFACE_ID;
+    }
+
+    /**
+    * @notice Execute the script as the Agent app
+    * @dev IForwarder interface conformance. Forwards any token holder action.
+    * @param _evmScript Script being executed
+    */
+    function forward(bytes _evmScript)
+        public
+        authP(RUN_SCRIPT_ROLE, arr(getScriptACLParam(_evmScript)))
+    {
+        bytes memory input = ""; // no input
+        address[] memory blacklist = new address[](0); // no addr blacklist, can interact with anything
+        runScript(_evmScript, input, blacklist);
+        // We don't need to emit an event here as EVMScriptRunner will emit ScriptResult if successful
+    }
+
+    function isValidSignature(bytes32 hash, bytes signature) public view returns (bytes4) {
+        // Short-circuit in case the hash was presigned. Optimization as performing calls
+        // and ecrecover is more expensive than an SLOAD.
+        if (isPresigned[hash]) {
```

Comment:
Referring to [this discussion](https://forum.aragon.org/t/agent-app-arbitrary-actions-from-daos/275)

1. > A standard for contracts ‘signing messages’ (already live in 0x v2 3), is for the contract to expose a isValidSignature function that gets called for verifying whether the contract approves a given signature as its own:

      /* nitpicking alert */: I am a bit unsure about this one. the method is called `isValidSignature` but the provided signature is in some cases actually never checked. Instead the method performs a static hash lookup assuring that a certain anonymous hash is known by the Agent. It is kind of equivalent to a signature check given that only a tightly controlled role can add hashes but also works as kind of a backdoor to 'sign' arbitrary hashes (even downstream in the name of potentially configured signature providers; hash lookup takes precedence). What is the use-case for presigned hashes?

2. there is no way to remove presigned hashes from the map which might be an issue if someone accidentally "signs" a hash. How likely is the case that someone accidentally adds a wrong hash? Is adding a hash done by user interaction or in an automated fashion (contract, ...)? What is the worst case scenario if this happens (redeploy app?)? From the discussion I learned that this mechanism is intended to be irrevocable mimicking ordinary ecdsa signatures. 

3.  > 4.1.0 Designated signer is not set
  Return false unless the hash has been pre-signed (see section 4.2)
  
    spec mismatch: With the current implementation `SignatureValidator.isValidSignature` is still being called to verify the signature even though `designatedSigner` might be addr(0).

Reply:
@izqui @sohkai 
`isValidSignature` function encapsulates the usage of two different mechanisms of signature validation. According to the [discussion](https://forum.aragon.org/t/agent-app-arbitrary-actions-from-daos/275), it’s intended behaviour that one of those mechanisms processes irrevocable signatures (presigned) and the other one works with signatures that can be revoked. The problem is that there is no well-defined way to know whether this particular signature is revocable or not. Different types of signatures represent different types of commitments, so they should often be used for different purposes and there should be a way of differing these types of signatures. So I would suggest 2 possible solutions here:
1. Keep single signature type and work with only revocable (or irrevocable) signatures.
2. Define a transparent mechanism that will distinguish revocable signatures from irrevocable.

P.S. Right now it's possible to find out if a signature is irrevocable by looking straight into public `isPresigned` storage variable, but I'm not sure if it's transparent enough. I think there should be an explicit function with clear naming for that.

P.P.S. Since this issue should be very common, we might also consider adding it to ERC1271